File: reference-implementations/dotnet/TomlSchema/SchemaExtractor.cs
namespace TomlSchema;

using System.Text;
using Tomlyn;
using Tomlyn.Model;

internal static class SchemaExtractor
{
    private const string CurrentSchemaVersion = "1.0.0";

    public static string Generate(TomlTable document)
    {
        var schema = new StringBuilder()
            .Append("[toml-schema]\n")
            .Append($"version = \"{CurrentSchemaVersion}\"\n\n")
            .Append("[elements]\n");
        foreach (var key in document.Keys.Order(StringComparer.Ordinal))
        {
            if (key != "toml-schema")
                AppendDefinition(schema, ["elements", key], document[key]);
        }
        return schema.ToString();
    }

    public static void ExtractFile(string documentPath, string schemaPath)
    {
        var document = TomlSerializer.Deserialize<TomlTable>(File.ReadAllText(documentPath))
            ?? throw new InvalidOperationException($"Failed to parse document: {documentPath}");
        File.WriteAllText(schemaPath, Generate(document), new UTF8Encoding(false));
    }

    private static void AppendDefinition(StringBuilder schema, IReadOnlyList<string> path, object? value)
    {
        schema.Append('\n').Append('[')
            .Append(string.Join(".", path.Select(EncodeKey)))
            .Append("]\n");
        var type = SchemaTypeOf(value);
        schema.Append("type = \"").Append(type).Append("\"\n");
        if (value is TomlArray array)
            schema.Append("itemtype = \"").Append(InferItemType(array.Cast<object?>())).Append("\"\n");
        else if (value is TomlTableArray tableArray)
            schema.Append("itemtype = \"").Append(InferItemType(tableArray.Cast<object?>())).Append("\"\n");
        if (value is TomlTable table)
        {
            foreach (var key in table.Keys.Order(StringComparer.Ordinal))
                AppendDefinition(schema, [.. path, key], table[key]);
        }
    }

    private static string InferItemType(IEnumerable<object?> items)
    {
        using var enumerator = items.GetEnumerator();
        if (!enumerator.MoveNext())
            return "any";
        var first = SchemaTypeOf(enumerator.Current);
        while (enumerator.MoveNext())
        {
            if (SchemaTypeOf(enumerator.Current) != first)
                return "any";
        }
        return first;
    }

    private static string SchemaTypeOf(object? value) => value switch
    {
        string => "string",
        long => "integer",
        double => "float",
        bool => "boolean",
        DateTime dt when dt.Kind == DateTimeKind.Utc => "offset-date-time",
        DateTime => "local-date-time",
        DateOnly => "local-date",
        TimeOnly => "local-time",
        TomlDateTime { Kind: TomlDateTimeKind.OffsetDateTimeByZ or TomlDateTimeKind.OffsetDateTimeByNumber } => "offset-date-time",
        TomlDateTime { Kind: TomlDateTimeKind.LocalDateTime } => "local-date-time",
        TomlDateTime { Kind: TomlDateTimeKind.LocalDate } => "local-date",
        TomlDateTime { Kind: TomlDateTimeKind.LocalTime } => "local-time",
        TomlArray or TomlTableArray => "array",
        TomlTable => "table",
        _ => "any"
    };

    private static string EncodeKey(string key)
    {
        if (key.Length > 0 && key.All(character =>
                character is >= 'A' and <= 'Z'
                    or >= 'a' and <= 'z'
                    or >= '0' and <= '9'
                    or '_' or '-'))
            return key;

        var encoded = new StringBuilder("\"");
        foreach (var character in key)
        {
            switch (character)
            {
                case '\\': encoded.Append("\\\\"); break;
                case '"': encoded.Append("\\\""); break;
                case '\b': encoded.Append("\\b"); break;
                case '\t': encoded.Append("\\t"); break;
                case '\n': encoded.Append("\\n"); break;
                case '\f': encoded.Append("\\f"); break;
                case '\r': encoded.Append("\\r"); break;
                default:
                    if (character < 0x20 || character == 0x7f)
                        encoded.Append($"\\u{(int)character:X4}");
                    else
                        encoded.Append(character);
                    break;
            }
        }
        return encoded.Append('"').ToString();
    }
}

File: reference-implementations/java/src/main/java/org/tomlschema/SchemaExtractor.java
package org.tomlschema;

import org.tomlj.TomlArray;
import org.tomlj.TomlTable;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

final class SchemaExtractor {
    private SchemaExtractor() {
    }

    static String generate(TomlTable document) {
        StringBuilder schema = new StringBuilder()
                .append("[toml-schema]\n")
                .append("version = \"").append(TomlSchemaVersion.CURRENT).append("\"\n\n")
                .append("[elements]\n");
        for (String key : sortedKeys(document)) {
            if (!key.equals("toml-schema")) {
                appendDefinition(schema, List.of("elements", key), document.get(List.of(key)));
            }
        }
        return schema.toString();
    }

    private static void appendDefinition(StringBuilder schema, List<String> path, Object value) {
        schema.append("\n[");
        for (int index = 0; index < path.size(); index++) {
            if (index > 0) {
                schema.append('.');
            }
            schema.append(encodeKey(path.get(index)));
        }
        String type = schemaType(value);
        schema.append("]\ntype = \"").append(type).append("\"\n");
        if (value instanceof TomlArray array) {
            schema.append("itemtype = \"").append(inferItemType(array)).append("\"\n");
        }
        if (value instanceof TomlTable table) {
            for (String key : sortedKeys(table)) {
                List<String> childPath = new ArrayList<>(path);
                childPath.add(key);
                appendDefinition(schema, childPath, table.get(List.of(key)));
            }
        }
    }

    private static String schemaType(Object value) {
        return switch (value) {
            case null -> "any";
            case String _ -> "string";
            case Long _ -> "integer";
            case Double _ -> "float";
            case Boolean _ -> "boolean";
            case OffsetDateTime _ -> "offset-date-time";
            case LocalDateTime _ -> "local-date-time";
            case LocalDate _ -> "local-date";
            case LocalTime _ -> "local-time";
            case TomlArray _ -> "array";
            case TomlTable _ -> "table";
            default -> "any";
        };
    }

    private static String inferItemType(TomlArray array) {
        if (array.isEmpty()) {
            return "any";
        }
        String first = schemaType(array.get(0));
        for (int index = 1; index < array.size(); index++) {
            if (!schemaType(array.get(index)).equals(first)) {
                return "any";
            }
        }
        return first;
    }

    private static List<String> sortedKeys(TomlTable table) {
        List<String> keys = new ArrayList<>(table.keySet());
        Collections.sort(keys);
        return keys;
    }

    private static String encodeKey(String key) {
        if (!key.isEmpty() && key.chars().allMatch(character ->
                Character.isLetterOrDigit(character) && character < 128
                        || character == '_' || character == '-')) {
            return key;
        }
        StringBuilder encoded = new StringBuilder("\"");
        for (int index = 0; index < key.length(); index++) {
            char character = key.charAt(index);
            switch (character) {
                case '\\' -> encoded.append("\\\\");
                case '"' -> encoded.append("\\\"");
                case '\b' -> encoded.append("\\b");
                case '\t' -> encoded.append("\\t");
                case '\n' -> encoded.append("\\n");
                case '\f' -> encoded.append("\\f");
                case '\r' -> encoded.append("\\r");
                default -> {
                    if (character < 0x20 || character == 0x7f) {
                        encoded.append(String.format("\\u%04X", (int) character));
                    } else {
                        encoded.append(character);
                    }
                }
            }
        }
        return encoded.append('"').toString();
    }
}
